includes/database.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace Wporg\TranslationEvents;
+
+use Exception;
+use WP_Query;
+use Wporg\TranslationEvents\Attendee\Attendee;
+
+class Database {
+	public static function upgrade(): void {
+		require_once ABSPATH . 'wp-admin/includes/upgrade.php';
+		global $gp_table_prefix;
+
+		// Create actions table.
+		$create_actions_table = "
+		CREATE TABLE `{$gp_table_prefix}event_actions` (
+			`translate_event_actions_id` int(11) unsigned NOT NULL AUTO_INCREMENT,
+			`event_id` int(10) NOT NULL COMMENT 'Post_ID of the translation_event post in the wp_posts table',
+			`original_id` int(10) NOT NULL COMMENT 'ID of the translation',
+			`user_id` int(10) NOT NULL COMMENT 'ID of the user who made the action',
+			`action` enum('approve','create','reject','request_changes') NOT NULL COMMENT 'The action that the user made (create, reject, etc)',
+			`locale` varchar(10) CHARACTER SET latin1 COLLATE latin1_swedish_ci NOT NULL COMMENT 'Locale of the translation',
+			`happened_at` datetime NOT NULL COMMENT 'When the action happened, in UTC',
+		PRIMARY KEY (`translate_event_actions_id`),
+		UNIQUE KEY `event_per_translated_original_per_user` (`event_id`,`locale`,`original_id`,`user_id`)
+		) COMMENT='Tracks translation actions that happened during a translation event'";
+		dbDelta( $create_actions_table );
+
+		// Create attendees table.
+		$create_attendees_table = "
+		CREATE TABLE `{$gp_table_prefix}event_attendees` (
+			`translate_event_attendees_id` int(11) unsigned NOT NULL AUTO_INCREMENT,
+			`event_id` int(10) NOT NULL COMMENT 'Post_ID of the translation_event post in the wp_posts table',
+			`user_id` int(10) NOT NULL COMMENT 'ID of the user who is attending the event',
+			`is_host` tinyint(1) default 0 not null comment 'Whether the user is a host of the event',
+			`is_contributor` tinyint(1) default 0 not null comment 'Whether the user created or reviewed translations during the event',
```

Comment:
I guess we could indeed calculate it from it the `actions` table, though I don't know if that could be an expensive query, given the `actions` table will likely become a very large table with time.

For example, when displaying the event details page, we would need to calculate it for each attendee of the event.

Concerning conflicting information, I think that would not happen, because you can't go back to not being a contributor. I mean, if you contributed to an event, it's not possible to "undo" that contribution, right?

So once `is_contributor` is set to `true`, it will always be `true`, I think there is no scenario where it would become `false` again.

Personally I like that the data of attendees would be decoupled from the `actions`. If you think we should change it, I can do it.

Reply:
My main concerns here are:
- If we have a bug in the code, we could have inconsistent information in the tables. E.g., translations for a user in the `event_actions` table and this user with `is_contributor=0` in the `event_ateendees` table. I could assume that this is unlikely.
- If we store this information in the `event_ateendees` table, each time a user makes a translation, we are going to need to query this table to know if this user is contributing to this event, so we need to do a lot of extra queries when the users translate strings. If we have this information in the `event_actions` table, we will only query it when we need to show info for an event, and the query is not expensive, because it only uses one table (will maybe a lot of registers). We could cache this query in the future. And as Alex suggested [here](https://github.com/WordPress/wporg-gp-translation-events/issues/196#issuecomment-2043171940), we could compute it once a day, with a cron, for the events that have finished the last day.

What do you think about?

Reply:
These are fair points 👍

I also thought further about this and another problem with having `is_contributor` in the attendees table is that if the user would un-attend the event, the `is_contributor` would be lost. So we would need to prevent attendees who contributed from un-attending, which is not something I'm certain we want to do.

I will change it to remove `is_contributor` from the attendees table.

Reply:
What about we call it `was_contributor` so that it's clear that this is supposed to be calculated after the event is over?

Reply:
The idea is that we can use the `event_attendees` table to generate stats without having to query the `event_actions` table.

Reply:
Given that these stats are not a feature we have right now, could we delay this decision until we implement those stats? At that point we can decide to introduce the `was_contributor` field, or to calculate those stats from both the `attendees` and `actions` tables.

Reply:
Ok! Better than delaying this more.

Reply:
Ok, we can remove the `is_contributor` and move on with this PR to approve it.

Reply:
I'm on this.